You can't add more than 9 resource packs

That is, you can add more than 9 resourcepacks, just enable them via the /rp enable command, but the interface does not provide that someone can hold more than 9 resourcepacks, which causes a funny glitch like in the screenshot below

![изображение](https://github.com/user-attachments/assets/5c8035b7-966d-4cc0-b2e1-33bee6211f33)

Below I will give the error log that occurs when running the /rp list command:

```
org.bukkit.command.CommandException: Unhandled exception executing command 'rp' in plugin SimpleResourcepack v2.6
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:47) ~[paper-mojangapi-1.21.1-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.command.brigadier.bukkit.BukkitCommandNode$BukkitBrigCommand.run(BukkitCommandNode.java:91) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at com.mojang.brigadier.context.ContextChain.runExecutable(ContextChain.java:73) ~[brigadier-1.3.10.jar:?]
        at net.minecraft.commands.execution.tasks.ExecuteCommand.execute(ExecuteCommand.java:31) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.commands.execution.tasks.ExecuteCommand.execute(ExecuteCommand.java:19) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.commands.execution.UnboundEntryAction.lambda$bind$0(UnboundEntryAction.java:8) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.commands.execution.CommandQueueEntry.execute(CommandQueueEntry.java:5) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.commands.execution.ExecutionContext.runCommandQueue(ExecutionContext.java:103) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.commands.Commands.executeCommandInContext(Commands.java:443) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.commands.Commands.performCommand(Commands.java:350) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.commands.Commands.performCommand(Commands.java:337) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.commands.Commands.performCommand(Commands.java:332) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.performUnsignedChatCommand(ServerGamePacketListenerImpl.java:2218) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.lambda$handleChatCommand$11(ServerGamePacketListenerImpl.java:2192) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.TickTask.run(TickTask.java:18) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:151) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1535) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:201) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:125) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1512) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1505) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:135) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.MinecraftServer.managedBlock(MinecraftServer.java:1464) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:1471) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1316) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:329) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 18 out of bounds for length 18
        at java.base/java.util.Arrays$ArrayList.set(Arrays.java:4271) ~[?:?]
        at net.minecraft.core.NonNullList.set(NonNullList.java:50) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at org.bukkit.craftbukkit.inventory.CraftInventoryCustom$MinecraftInventory.setItem(CraftInventoryCustom.java:167) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at org.bukkit.craftbukkit.inventory.CraftInventory.setItem(CraftInventory.java:100) ~[paper-1.21.1.jar:1.21.1-131-84281ce]
        at SimpleResourcepack-2.6-1734716998249.jar/dev.thesheep.simpleresourcepack.api.ResourcepackGUIGenerator.openForPlayer(ResourcepackGUIGenerator.java:105) ~[SimpleResourcepack-2.6-1734716998249.jar:?]
        at SimpleResourcepack-2.6-1734716998249.jar/dev.thesheep.simpleresourcepack.api.subcommands.ListSubCommand.openForSelf(ListSubCommand.java:44) ~[SimpleResourcepack-2.6-1734716998249.jar:?]
        at SimpleResourcepack-2.6-1734716998249.jar/dev.thesheep.simpleresourcepack.api.subcommands.ListSubCommand.onSubCommand(ListSubCommand.java:15) ~[SimpleResourcepack-2.6-1734716998249.jar:?]
        at SimpleResourcepack-2.6-1734716998249.jar/dev.thesheep.simpleresourcepack.api.ResourcepackCommand.onCommand(ResourcepackCommand.java:63) ~[SimpleResourcepack-2.6-1734716998249.jar:?]
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[paper-mojangapi-1.21.1-R0.1-SNAPSHOT.jar:?]
        ... 27 more
```

--- a/.gitignore
+++ b/.gitignore
@@ -22,8 +22,9 @@ hs_err_pid*
 # Mac
 .DS_Store
 
-# Maven
-log/
+# Gradle
+gradlew.bat
+gradlew
 
 # Inclusions
 !lib/**

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -4,7 +4,7 @@ plugins {
 }
 
 group = "dev.thesheep"
-version = "2.6"
+version = "2.7"
 
 val earliestSupportedAPIVersion = "1.16"
 

--- a/src/main/java/dev/thesheep/simpleresourcepack/api/ResourcepackGUIGenerator.java
+++ b/src/main/java/dev/thesheep/simpleresourcepack/api/ResourcepackGUIGenerator.java
@@ -17,6 +17,8 @@ import java.util.ArrayList;
 import java.util.List;
 
 public class ResourcepackGUIGenerator {
+    private final List<Player> activeForPlayers = new ArrayList<>();
+
     public void openForPlayer(Player player)
     {
         // Make sure the player does not have any other guis open.
@@ -24,90 +26,58 @@ public class ResourcepackGUIGenerator {
 
         activeForPlayers.add(player);
 
-        Inventory inventory = Bukkit.createInventory(player, 9 * 2, "Resourcepacks");
+        SimpleResourcepack srp = SimpleResourcepack.getInstance();
+        List<String> resourcepacks = srp.getResourcepacks();
+        int resourcepackCount = resourcepacks.size();
+
+        int rowCount = ((resourcepackCount + 8) / 9) * 2;
+        if (rowCount > 6)
+        {
+            // TODO: Add pagination for support for > 27 resourcepacks.
+            throw new RuntimeException("Number of resourcepacks exceeds Paper's max inventory!");
+        }
+
+        Inventory inventory = Bukkit.createInventory(player, 9 * rowCount, "Resourcepacks");
         player.openInventory(inventory);
 
-        SimpleResourcepack srp = SimpleResourcepack.getInstance();
-
         int index = 0;
-        for(String foldername : srp.getResourcepacks())
+        for(String resourcepack : resourcepacks)
         {
-            File f = new File(srp.getSettingsFolder().toPath() + "/" + foldername + ".yml");
-
-            if(Files.exists(f.toPath()))
+            if (index % 9 == 0 && index != 0)
             {
-                YamlConfiguration configuration = YamlConfiguration.loadConfiguration(f);
-
-                String permission = configuration.getString("permission", "none");
-
-                if(permission.equalsIgnoreCase("none") || player.hasPermission(permission)) {
-
-                    Material material = Material.getMaterial(configuration.getString("material", "PAPER"));
-                    if(material == null)
-                    {
-                        Bukkit.getLogger().severe("Invalid material for pack " + foldername + "\"" + configuration.getString("material", "PAPER") + "\"");
-                        material = Material.PAPER;
-                    }
-                    ItemStack button = new ItemStack(material);
-                    ItemMeta meta = button.getItemMeta();
-                    meta.setDisplayName(configuration.getString("name", foldername));
-
-                    List<String> lore = configuration.getStringList("lore");
-                    meta.setLore(lore);
-
-                    button.setItemMeta(meta);
-                    inventory.setItem(index, button);
-
-                    List<String> defaults = SimpleResourcepack.getInstance().getConfig().getStringList("default");
-
-                    ItemStack state = new ItemStack(Material.BARRIER);
-                    if(defaults.contains(foldername))
-                    {
-                        state = new ItemStack(Material.GRAY_STAINED_GLASS_PANE);
-                        ItemMeta meta2 = state.getItemMeta();
-                        List<String> lore2 = new ArrayList<>();
-                        lore2.add("§fThis resourcepack is forced.");
-                        meta2.setDisplayName("§7§lDEFAULT");
-                        meta2.setLore(lore2);
-                        state.setItemMeta(meta2);
-                    }
-                    else {
-                        if (SimpleResourcepack.getInstance().getPlayerPref().getResourcepackPreferences(player).contains(foldername)) {
-                            state = new ItemStack(Material.GREEN_STAINED_GLASS_PANE);
-                            ItemMeta meta2 = state.getItemMeta();
-                            List<String> lore2 = new ArrayList<>();
-                            meta2.setDisplayName("§a§lENABLED");
-                            lore2.add("§fClick to disable.");
-                            meta2.setLore(lore2);
-
-                            NamespacedKey key = NamespacedKey.fromString("button_disable", SimpleResourcepack.getInstance());
-                            meta2.getPersistentDataContainer().set(key, PersistentDataType.STRING, foldername);
-
-                            state.setItemMeta(meta2);
-                        }
-                        else {
-                            state = new ItemStack(Material.RED_STAINED_GLASS_PANE);
-                            ItemMeta meta2 = state.getItemMeta();
-
-                            List<String> lore2 = new ArrayList<>();
-                            lore2.add("§fClick to enable.");
-
-                            meta2.setDisplayName("§c§lDISABLED");
-
-                            NamespacedKey key = NamespacedKey.fromString("button_enable", SimpleResourcepack.getInstance());
-                            meta2.getPersistentDataContainer().set(key, PersistentDataType.STRING, foldername);
-
-                            meta2.setLore(lore2);
-                            state.setItemMeta(meta2);
-                        }
-                    }
-
-                    inventory.setItem(index + 9, state);
-                }
+                index += 9;
             }
 
-            index++;
+            File f = new File(srp.getSettingsFolder().toPath() + "/" + resourcepack + ".yml");
+            if(!Files.exists(f.toPath()))
+            {
+                continue;
+            }
 
+            YamlConfiguration resourcepackData = YamlConfiguration.loadConfiguration(f);
+            String resourcepackPermission = resourcepackData.getString("permission", "none");
+
+            if(resourcepackPermission.equalsIgnoreCase("none") || player.hasPermission(resourcepackPermission))
+            {
+                Material material = Material.getMaterial(resourcepackData.getString("material", "PAPER"));
+                if(material == null)
+                {
+                    Bukkit.getLogger().severe("Invalid material for pack " + resourcepack + "\"" + resourcepackData.getString("material", "") + "\"");
+                    material = Material.PAPER;
+                }
+
+                ItemStack itemButton = new ItemStack(material);
+                ItemMeta meta = itemButton.getItemMeta();
+                meta.setDisplayName(resourcepackData.getString("name", resourcepack));
+                meta.setLore(resourcepackData.getStringList("lore"));
+
+                itemButton.setItemMeta(meta);
+                inventory.setItem(index, itemButton);
+
+                displayState(player, inventory, resourcepack, index + 9);
+            }
+
+            ++index;
         }
     }
 
@@ -115,11 +85,64 @@ public class ResourcepackGUIGenerator {
     {
         return activeForPlayers.contains(player);
     }
-    private List<Player> activeForPlayers = new ArrayList<>();
 
     public void closeForPlayer(Player player)
     {
         activeForPlayers.remove(player);
         player.closeInventory();
+    }
+
+    private void displayState(Player player, Inventory inventory, String resourcepack, int index)
+    {
+        List<String> defaults = SimpleResourcepack.getInstance().getConfig().getStringList("default");
+
+        ItemStack state;
+        if(defaults.contains(resourcepack))
+        {
+            state = new ItemStack(Material.GRAY_STAINED_GLASS_PANE);
+            ItemMeta meta = state.getItemMeta();
+
+            meta.setDisplayName("§7§lDEFAULT");
+
+            List<String> lore = new ArrayList<>(1);
+            lore.add("§fThis resourcepack is forced.");
+            meta.setLore(lore);
+
+            state.setItemMeta(meta);
+        }
+        else {
+            if (SimpleResourcepack.getInstance().getPlayerPref().getResourcepackPreferences(player).contains(resourcepack)) {
+                state = new ItemStack(Material.GREEN_STAINED_GLASS_PANE);
+                ItemMeta meta = state.getItemMeta();
+
+                meta.setDisplayName("§a§lENABLED");
+
+                List<String> lore = new ArrayList<>(1);
+                lore.add("§fClick to disable.");
+                meta.setLore(lore);
+
+                NamespacedKey key = NamespacedKey.fromString("button_disable", SimpleResourcepack.getInstance());
+                meta.getPersistentDataContainer().set(key, PersistentDataType.STRING, resourcepack);
+
+                state.setItemMeta(meta);
+            }
+            else {
+                state = new ItemStack(Material.RED_STAINED_GLASS_PANE);
+                ItemMeta meta2 = state.getItemMeta();
+
+                List<String> lore2 = new ArrayList<>();
+                lore2.add("§fClick to enable.");
+
+                meta2.setDisplayName("§c§lDISABLED");
+
+                NamespacedKey key = NamespacedKey.fromString("button_enable", SimpleResourcepack.getInstance());
+                meta2.getPersistentDataContainer().set(key, PersistentDataType.STRING, resourcepack);
+
+                meta2.setLore(lore2);
+                state.setItemMeta(meta2);
+            }
+        }
+
+        inventory.setItem(index, state);
     }
 }

--- a/src/main/java/dev/thesheep/simpleresourcepack/api/subcommands/DisableSubCommand.java
+++ b/src/main/java/dev/thesheep/simpleresourcepack/api/subcommands/DisableSubCommand.java
@@ -48,7 +48,7 @@ public class DisableSubCommand extends SubCommand {
         Configuration configuration = SimpleResourcepack.getInstance().getResourcepackConfig(packName);
         if (configuration == null)
         {
-            commandSender.sendMessage("§cResourcepack settings where not found.");
+            commandSender.sendMessage("§cResourcepack settings were not found.");
             return;
         }
 

--- a/src/main/java/dev/thesheep/simpleresourcepack/api/subcommands/EnableSubCommand.java
+++ b/src/main/java/dev/thesheep/simpleresourcepack/api/subcommands/EnableSubCommand.java
@@ -48,7 +48,7 @@ public class EnableSubCommand extends SubCommand {
         Configuration configuration = SimpleResourcepack.getInstance().getResourcepackConfig(packName);
         if (configuration == null)
         {
-            commandSender.sendMessage("§cResourcepack settings where not found.");
+            commandSender.sendMessage("§cResourcepack settings were not found.");
             return;
         }
 

